src/utils.jl

```diff
@@ -44,19 +44,83 @@ vcat_arrays_vector(arrays_vector::AbstractVector{Param}) = vcat(Knet.cat1d.(arra
 """
     reset_minibatch_train!(nlp::AbstractKnetNLPModel)
 
-Select a new training minibatch for `nlp`.
-Typically used before a new evaluation of the loss function/gradient.
+Select the first training minibatch for `nlp`.
 """
-reset_minibatch_train!(nlp::AbstractKnetNLPModel) =
-  nlp.current_training_minibatch = rand(nlp.training_minibatch_iterator)
+function reset_minibatch_train!(nlp::AbstractKnetNLPModel)
+  nlp.current_training_minibatch = first(nlp.training_minibatch_iterator)
+  nlp.i_train = 1
+end
+
+"""
+  rand_minibatch_train!(nlp::AbstractKnetNLPModel)
+
+Select a training minibatch for `nlp` randomly.
+"""
+function rand_minibatch_train!(nlp::AbstractKnetNLPModel)
+  nlp.i_train = rand(1:(nlp.training_minibatch_iterator.imax))
+  nlp.current_training_minibatch = iterate(nlp.training_minibatch_iterator, nlp.i_train)
+end
+
+"""
+    minibatch_next_train!(nlp::AbstractKnetNLPModel)
+
+Selects the next mini-batch from `nlp.training_minibatch_iterator`.  
+Returns the new current location of the iterator `nlp.i_train`.
+If it returns 1, the current training minibatch is the first of `nlp.training_minibatch_iterator` and the previous minibatch was the last of `nlp.training_minibatch_iterator`.
+`minibatch_next_train!` aims to be used in a loop or method call - refere to KnetNLPModelProblems.jl for more use cases
+"""
+function minibatch_next_train!(nlp::AbstractKnetNLPModel)
+  nlp.i_train += nlp.size_minibatch # update the i by mini_batch size
+  if (nlp.i_train >= nlp.training_minibatch_iterator.imax)
+    # reset to the begining 
+    nlp.current_training_minibatch = first(nlp.training_minibatch_iterator) # reset to the first one
+    nlp.i_train = 1
+  else
```

Comment:
why not use `try ... catch` to reset to `first` when `iterate` returns `nothing`?

Reply:
I suggest we put this in a if else since the iterator function returns nothing if out of bound.
I think if else is less memory heavy than catch-try, no?
but I changed the code to be more robust

Reply:
@dpo I understoof from my course that Exceptions must be avoid, to not slow down the execution with system calls?
It is implemented differently in julia?

Reply:
Actually you wouldn't need a `try ... catch` because `iterate` returns `nothing` (it doesn't raise an exception). It seems strange to me to index an iterator.

Reply:
exactly what I thought first I used it 
but Knet uses this code for iterate: https://github.com/denizyuret/Knet.jl/blob/master/src/train20/data.jl not sure if this is correct or not 
Here is their iterated:
`
@propagate_inbounds function iterate(d::Data, i=0)     # returns data in d.indices[i+1:i+batchsize]
    if i >= d.imax
        return nothing
    end
    if d.shuffle && i == 0
        d.indices = randperm(d.length)
    end
    nexti = min(i + d.batchsize, d.length)
    ids = d.indices[i+1:nexti]
    xbatch = try convert(d.xtype, reshape(d.x[:,ids],d.xsize[1:end-1]...,length(ids)))
    catch e; throw(DimensionMismatch("Error converting batch from $(summary(d.x)) to $(d.xtype): $e")); end
    if d.y == nothing
        return (xbatch,nexti)
    else
        ybatch = try convert(d.ytype, reshape(d.y[:,ids],d.ysize[1:end-1]...,length(ids)))
        catch; throw(DimensionMismatch("Y tensor not compatible with size=$(d.ysize) and type=$(d.ytype)")); end
        return ((xbatch,ybatch),nexti)
    end
end


@propagate_inbounds function iterate(r::Repeat, s=(1,))
    epoch, state = s[1], tail(s)
    epoch > r.n && return nothing
    next = iterate(r.data, state...)
    next === nothing && return iterate(r, (epoch+1,))
    (next[1], (epoch, next[2]))
end


`

Reply:
@dpo  I updated the code or put comments that needs your input, let me know what you think